src/main/java/com/hms/flexyosisoftconnector/OSIsoftConnectorMain.java

```diff
@@ -170,78 +179,105 @@ private static void getNewPayloadToSend() {
+          Logger.LOG_EXCEPTION(e);
+          Logger.LOG_WARN("There was an unknown error collecting data points!");
         }
 
-        Logger.LOG_DEBUG("Grabbed " + queuePoints.size() + " datapoints for proccessing.");
-        for (int i = 0; i < queuePoints.size(); i++) {
-          DataPoint data = ((DataPoint) queuePoints.get(i));
+        // if there are data points to send
+        if (queuePoints.size() != 0) {
+          // Check if queue is behind
+          try {
+            long queueBehindMillis = HistoricalDataQueueManager.getQueueTimeBehindMillis();
+            if (queueBehindMillis >= OSIsoftConfig.WARNING_LIMIT_QUEUE_BEHIND_MILLISECONDS) {
+              Logger.LOG_WARN(
+                  "The historical data queue is running behind by "
+                      + SCTimeUtils.getDayHourMinSecsForMillis(queueBehindMillis));
+            }
+          } catch (IOException e) {
+            Logger.LOG_SERIOUS("Unable to detect if historical data queue is running behind.");
+            Logger.LOG_EXCEPTION(e);
+          } catch (TimeTrackerUnrecoverableException e) {
+            Logger.LOG_SERIOUS("Time tracker file has been corrupted.");
+            Logger.LOG_EXCEPTION(e);
+          }
 
-          // build a payload out of data points
-          if (data != null) {
-            payloadMngr.appendDataPointToPayLoad(data);
+          Logger.LOG_DEBUG("Grabbed " + queuePoints.size() + " datapoints for proccessing.");
+          for (int i = 0; i < queuePoints.size(); i++) {
+            DataPoint data = ((DataPoint) queuePoints.get(i));
+
+            // build a payload out of data points
+            if (data != null) {
+              payloadMngr.appendDataPointToPayLoad(data);
+            }
+
+            // Sleep between datapoints to significantly increase Flexy performance.
+            Thread.yield();
+            try {
+              final int threadWaitMS = 5;
+              Thread.sleep(threadWaitMS);
+            } catch (InterruptedException e) {
+              Logger.LOG_EXCEPTION(e);
+              Logger.LOG_SERIOUS("Exception thrown while sleeping thread.");
+            }
           }
+        }
+      } else {
+        payloadQueueFullCount += 1;
 
-          // Sleep between datapoints to significantly increase Flexy performance.
-          Thread.yield();
-          try {
-            final int threadWaitMS = 5;
-            Thread.sleep(threadWaitMS);
-          } catch (InterruptedException e) {
-            Logger.LOG_EXCEPTION(e);
-            Logger.LOG_SERIOUS("Exception thrown while sleeping thread.");
+        if (payloadQueueFullCount >= payloadQueueFullWarnThresh) {
+          payloadQueueFullCount = 0;
+          Logger.LOG_DEBUG("Reading historical data skipped due to full payload queue.");
+
+          // after warning increase the threshold, but avoid roll over
+          if (payloadQueueFullWarnThresh < Long.MAX_VALUE / 100) {
+            payloadQueueFullWarnThresh *= 10;

```

Comment:
Magic numbers. 

Also, why `/ 100` and why ` *= 10`? This code is not very understandable on its own + as-is. I think an updated comment will also help here.

Reply:
I have not addressed this comment

Reply:
will add a comment